src/connectWithShell.tsx

```diff
@@ -140,7 +144,7 @@ export interface ConnectWithShellOptions {
     /**
      * Update the component only when this function returns true
      */
-    shouldComponentUpdate?(shell: Shell): boolean
+    shouldComponentUpdate?(shell: Shell, nextProps?: OwnProps | StateProps): boolean
```

Comment:
I noticed it, this is why the interface looks like this `shouldComponentUpdate(shell: Shell, nextProps?: OwnProps | StateProps)`. `nextProps` may be `OwnProps` or `StateProps` as `shouldComponentUpdate` is called in 3 places:
1. wrapping `mapStateToProps`(memoWithShouldUpdate) - will receive `ownProps`
2. `areOwnPropsEqual ` -  will receive `ownProps`
3. `areStatePropsEqual` - in this case it will receive next state props

I've done it like that in order to cover all possible usage.

But I agree that 3nd option(stateProps) doesn't make a lot of sense, so we can go only with `ownProps`: 1st and 2d option described above.

So, in the end we can go with `shouldComponentUpdate(shell: Shell, ownProps?: OwnProps)` which will cover most of the cases, WDYT?

Reply:
It might make sense to proceed with `shouldComponentUpdate(shell: Shell, ownProps?: OwnProps)`, but we should go through all the use cases thoroughly.

In case of `memoWithShouldUpdate` over `mapStateToProps` - we can pass `OwnProps` - would work as expected.
In case of `areOwnPropsEqual` - we can pass `OwnProps` - would work as expected.
In case of `areStatePropsEqual` - we will have to call `shouldComponentUpdate` **without any props**, because we only have access to next / prev state props here - might result in weird behaviors.

So for the `areStatePropsEqual` case - there are 2 interesting cases when `OwnProps` are not passed:

A. `shouldComponentUpdate`, should be false but returned true 
(e.g. `if (ownProps?.shouldOptimize) { return false }`) -> 
`areStatePropsEqual` props will redundantly be compared, but with memoized props so it's always equal - undesirable but we may be able to live with it.

B. `shouldComponentUpdate`, should be true but returned false 
(e.g. `if (ownProps?.shouldSkipOptimization) { return true }`) -> 
`areStatePropsEqual` props will not be compared  - even though `mapStateToProps` will run again and return new props - so no update, even though update should happen, this is a bug, unless I'm missing something.

That is, unless we can find some way to inject `OwnProps` to `areStatePropsEqual`

Any idea?

Reply:
We are tighten by initial implementation of `shouldComponentUpdate` that are used in all 3 cases. My suggestion to go only with `ownProps` and pass `undefined` in case of `areStatePropsEqual`. I agree with your thought about different behaviour in those casees, let's try to go 1 by 1:

A. We will have some redundant comparison, but in case of workaround(using `key` to force render) we will have same amount of work but with much worse DX: key hack vs intended interface

B. This one, looks like may lead to bugs. After checking [react-redux codebase](https://github.com/reduxjs/react-redux/blob/master/src/connect/selectorFactory.ts#L145) it may be ok, due to order of `areStatePropsEqual`/`areOwnPropsEqual` execution. But I'm not sure before I'll deep dive into internals of react-redux. And depend on internal react-redux implementation, probably not a good idea.

What if we can use `this.props` of `ConnectedComponent` class which is actually `ownProps` of [this.connectedComponent](https://github.com/wix-incubator/repluggable/blob/6e51a68dc4a28d0d633be91961e86354e31bd9a7/src/connectWithShell.tsx#L90) and pass it `shouldComponentUpdate` in all 3 cases. It should work as `this.props` will be updated before `this.connectedComponent` is rendered.